### ISSUE TITLE
do not report too-few-public-methods for `attr.s`, `attr.dataclass`, `typing.NamedTuple` and similar. 

### ISSUE THREAD

**Issue by graingert** [pylint-dev/pylint#3732]
do not report too-few-public-methods for `attr.s`, `attr.dataclass`, `typing.NamedTuple` and similar. 

example:

```python
import attr

@attr.dataclass(frozen=True)  # pylint-ignore: too-few-public-methods
class Foo:
    ham: str
    spam: str
```

> @uriva As mentioned by @matkoniecz in most cases the check makes sense. If your class has a single method on it, it's probably a function in disguise. You can also configure the number of public methods with [min-public-methods](http://pylint.pycqa.org/en/latest/technical_reference/features.html#design-checker-options). But there are a couple of classes for which we should not report this, as it makes perfect sense to not have too many methods or no methods at all. This includes `attrs`, `dataclasses`, `typing.NamedTuple` and similar. 

_Originally posted by @PCManticore in https://github.com/PyCQA/pylint/issues/2710#issuecomment-458157179_

**Comment by uriva** [comment]
I'm not sure why this reopened, I am no longer using pylint (moved to flake8) so no interest in this anymore.
Thanks

**Comment by matkoniecz** [comment]
I am also not sure why you copied the comment outright pinging all involved rather than posting only relevant part.

Posting "I'm going to close this issue" while opening an issue is bizarre.

**Comment by graingert** [comment]
@matkoniecz comment now quoted, code example added

**Comment by graingert** [comment]
> I am also not sure why you copied the comment outright pinging all involved rather than posting only relevant part.

I just hit the "reference in new issue button"

**Comment by Pierre-Sassoulas** [comment]
This message is probably going to be disabled by default because there is a lot of example of place where it makes sense to have few public method (Django Meta class, inheritance of class that implements a run() function etc.) See #3512 . If you think this is useful and want to use it, but with a whitelist of class that can have few public method, I don't think there is an easy way to do that right now, but I think this is doable.

**Comment by earonesty** [comment]
Configurable suppressions for classes would be generally useful:

Simple config override in .pylintrc for classes that:

# [CLASSRULES.<CUSTOM_NAME>]
# metaclass_match=<metaclass_regex>
# parent_match=<parent_regex>
# disable=list,of,disables

Would solve a lot of these issues.

**Comment by Pierre-Sassoulas** [comment]
Also true for exceptions see : #4464

**Comment by wnoise** [comment]
Also true for anything inheriting from typing.Protocol.

**Comment by graingert** [comment]
@Pierre-Sassoulas why was this unpinned?

**Comment by graingert** [comment]
this should also not report: `too-many-instance-attributes` for attr.s

**Comment by Pierre-Sassoulas** [comment]
I unpinned it because there was more pressing issues to advertise first (ranked according to number of thumbsup or heart reactions), it does not mean that I would not review a fix for this ;)

**Comment by g1itch** [comment]
What about context managers? May pylint detect them by `__enter__()` and `__exit__()` methods defined?

**Comment by davebirch** [comment]
If this is still for consideration, attrs now support an `@define` decorator, as well as `@attr.s`.

**Comment by emirkmo** [comment]
please also Add Protocol classes. I just want pylint to ignore my Protocol classes so I don't have to make 10 line protocol classes just to make pylint happy for no reason. I want to use them for typing.
